### DISCUSSION TITLE
Taking subsequence of tree sequence object

### DISCUSSION THREAD

**Discussion by mshpak76** [tskit-dev/tskit#2753] [Q&A] [answered]
For reasons specific to issues I've encountered with the recombination and gene conversion parameters that I'm working with, I would like to use tskit tools to do the following:

Suppose that I've generated a tree sequence object where initializeGenomicEelment() is given a length of 15kb. Given the output tree, I would like to create a new tree sequence object using only the middle 5 kb but retaining the tree structure.

Is there a simple tskit function that will let me take a ts object and specify a subset/range of sites to consider?

Thank you for any suggestions.

**Comment by molpopgen** [comment]
[This](https://tskit.dev/tskit/docs/stable/python-api.html?highlight=keep_interval#tskit.TreeSequence.keep_intervals) may be what you are looking for.

**Reply by mshpak76** [reply]
Thanks. If I simulated a 15 kb sequence and generated tree sequence ts, I assume that

     \mts =ts.keep_intervals([[4999,9999]])

should give me the middle 5 kb interval, correct? I applied this function to my ts, and when I ran segregating_sites() on the object, I obtained a value error

    ValueError: Sample sets must contain at least one element

It seems implausible to me that given my parameters none of the segregating sites would be in the middle 1/3 of the sequence, since segregating_sites() on the entire sequence always returns a large number of simulated SNPs. Is there some issue with how I'm applying keep_intervals() ?

**Reply by molpopgen** [reply]
Hard to know without knowing more about your work flow.  Simple examples work as expected:

```
>>> ts = msprime.sim_ancestry(200, sequence_length=1000, recombination_rate=1e-4)
>>> len([e for e in ts.edges() if e.left < 500 or e.left > 600])
802
>>> x = ts.keep_intervals([(500,600)])
>>> len([e for e in x.edges() if e.left < 500 or e.left > 600])
0
```

**Reply by jeromekelleher** [reply]
It sounds like you're getting an error with how you're calling ``segregating_sites`` rather than from the tree sequence operations @mshpak76 - can you paste in how you're calling it here (ideally with a minimum example based on msprime simulations that we can reproduce directly)?

**Reply by mshpak76** [reply]
I will try to rewrite my code to create a short, self-contained "toy" example that reproduces the problem. In the meantime, here is the recapitation code that I have that imports a 3-population ts object from slim, followed by the segregating sites call. As noted before, segregating_sites() works fine and generates output until I introduced the keep_intervals() function:

    ts = tskit.load("ThreePopFix.trees")

    ts_new = pyslim.SlimTreeSequence(ts)

    num_replicates = 1
    make_seed = random.randint(0,2**31)
    rng = np.random.RandomState(make_seed)
    recap_seed = rng.randint(1,2**31)
    ancestry_seed = rng.randint(1,2**31)  
    mutation_seed = rng.randint(1,2*31)

    rts = pyslim.recapitate(ts_new, recombination_rate = 1.09e-6, gene_conversion_rate = 6.25e-6, gene_conversion_tract_length = 518, ancestral_Ne=2e4, random_seed = ancestry_seed)

    mts = msprime.sim_mutations(rts, rate = 5.21e-7, keep=True, random_seed = mutation_seed, model=msprime.SLiMMutationModel(type=0))

    mts = mts.keep_intervals([[4999,9999]])
    outfile = open("slim_sweep_rep_output_dros", "w")

    samp_size = 25
    hap_size = samp_size * 2
    seg_sites = mts.segregating_sites(sample_sets=[mts.samples(population=1), mts.samples(population=2), mts.samples(population=3),]).tolist()

The last seg_sites call is what throws the error when combined with the keep_intervals() function call

**Reply by jeromekelleher** [reply, marked as answer]
Ah, you should probably add ``simplify=False`` to keep_intervals. Otherwise I think your populations may be renumbered as a result.

This is quite confusing behaviour, I think you're not the first person to get tripped up on it! I seem to remember we had a discussion about this recently, but can't find it.

**Reply by mshpak76** [reply]
I just tested the code with the simplify=False argument added to keep_intervals() and it no longer throws the error.
Thanks, Max

**Reply by petrelharp** [reply]
Hm, yes - we probably should have made "don't reorder/remove populations" the default behaviour basically everywhere.
